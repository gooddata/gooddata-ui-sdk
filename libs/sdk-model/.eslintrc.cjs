// (C) 2020 GoodData Corporation
module.exports = {
    parser: "@typescript-eslint/parser",
    plugins: ["prettier", "sonarjs", "eslint-plugin-tsdoc", "regexp"],
    extends: [
        "@gooddata",
        "plugin:import-esm/recommended",
        "plugin:sonarjs/recommended",
        "plugin:regexp/recommended",
        "../../.eslintrc.js",
    ],
    parserOptions: { tsconfigRootDir: __dirname },
};
