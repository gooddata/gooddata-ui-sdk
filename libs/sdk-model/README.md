# GoodData.UI domain models

[![npm version](https://img.shields.io/npm/v/@gooddata/sdk-model)](https://www.npmjs.com/@gooddata/sdk-model)&nbsp;
[![npm monthly downloads](https://img.shields.io/npm/dm/@gooddata/sdk-model)](https://npmcharts.com/compare/@gooddata/sdk-model?minimal=true)&nbsp;
![typescript](https://img.shields.io/badge/typescript-first-blue?logo=typescript)

This package is a part of the [GoodData.UI SDK](https://sdk.gooddata.com/gooddata-ui/docs/about_gooddataui.html).
To learn more, check [the source monorepo](https://github.com/gooddata/gooddata-ui-sdk).

This package provides domain models for GoodData.UI.

These domain models are backend-agnostic. This makes them reusable across different [Analytical Backend](https://www.npmjs.com/package/@gooddata/sdk-backend-spi) implementations.
The package includes TypeScript type definitions, factory functions, functions to get or set certain properties
of the objects in an immutable way, and more. These are used in both the `@gooddata/sdk-backend-*` and `@gooddata/sdk-ui* packages`.

## License

(C) 2017-2022 GoodData Corporation

This project is under MIT License. See [LICENSE](https://github.com/gooddata/gooddata-ui-sdk/blob/master/libs/sdk-model/LICENSE).
