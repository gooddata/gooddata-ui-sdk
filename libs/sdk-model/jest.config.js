// (C) 2019 GoodData Corporation
import base from "../../common/config/jest/jest.config.base.mjs";

export default {
  ...base,
};