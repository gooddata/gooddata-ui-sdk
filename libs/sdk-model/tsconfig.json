{
    "compilerOptions": {
        "target": "es2017",
        "lib": ["es2017", "dom"],
        "module": "NodeNext",
        "moduleResolution": "nodenext",
        "esModuleInterop": true,
        "importHelpers": true,        
        "sourceMap": true,
        "outDir": "./esm",
        "strict": true,
        "noImplicitAny": true,
        "noUnusedLocals": true,
        "noUnusedParameters": true,
        "noImplicitReturns": true,
        "removeComments": false
    },

    "exclude": ["**/node_modules","esm"]
}
