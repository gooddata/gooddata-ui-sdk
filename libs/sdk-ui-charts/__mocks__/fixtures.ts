// (C) 2007-2019 GoodData Corporation
import { LegacyExecutionRecording, legacyRecordedDataView } from "@gooddata/sdk-backend-mockingbird";

import * as rec from "./recordings/playlist.js";
import { DataViewFacade } from "@gooddata/sdk-ui";

function legacyRecordedDataFacade(rec: LegacyExecutionRecording): DataViewFacade {
    return DataViewFacade.for(legacyRecordedDataView(rec));
}

//
// new fixtures
//

export const testWorkspace = "testWorkspace";

//
// Area chart fixtures
//

export const areaChartWith3MetricsAndViewByAttribute = legacyRecordedDataFacade(
    rec.AreaChartWith3MetricsAndViewByAttribute,
);

export const areaChartWithMeasureViewByAndStackBy = legacyRecordedDataFacade(
    rec.AreaChartWithMeasureViewByAndStackBy,
);

//
// Bar chart fixtures
//

export const barChartWith4MetricsAndViewByTwoAttributes = legacyRecordedDataFacade(
    rec.BarChartWith4MetricsAndViewByTwoAttributes,
);
export const barChartWithStackByAndViewByAttributes = legacyRecordedDataFacade(
    rec.BarChartWithStackByAndViewByAttributes,
);
export const barChartWithViewByAttribute = legacyRecordedDataFacade(rec.BarChartWithViewByAttribute);

export const barChartWith3MetricsAndViewByAttribute = legacyRecordedDataFacade(
    rec.BarChartWith3MetricsAndViewByAttribute,
);
export const barChartWith3MetricsAndViewByAttributeFunformat = legacyRecordedDataFacade(
    rec.BarChartWith3MetricsAndViewByAttributeFunformat,
);

export const barChartWith3MetricsAndViewByAttributePercInFormat = legacyRecordedDataFacade(
    rec.BarChartWith3MetricsAndViewByAttributePercInFormat,
);

export const barChartWithPopMeasureAndViewByAttribute = legacyRecordedDataFacade(
    rec.BarChartWithPopMeasureAndViewByAttribute,
);

export const barChartWithPopMeasureAndViewByAttributeX6 = legacyRecordedDataFacade(
    rec.BarChartWithPopMeasureAndViewByAttributeX6,
);

export const barChartWithPreviousPeriodMeasure = legacyRecordedDataFacade(
    rec.BarChartWithPreviousPeriodMeasure,
);
export const barChartWithSingleMeasureAndNoAttributes = legacyRecordedDataFacade(
    rec.BarChartWithSingleMeasureAndNoAttributes,
);

export const barChartWithPreviousPeriodMeasureX6 = legacyRecordedDataFacade(
    rec.BarChartWithPreviousPeriodMeasureX6,
);

//
// Bubble chart
//
export const bubbleChartWith2MetricsAndAttributeNoPrimaries = legacyRecordedDataFacade(
    rec.BubbleChartWith2MetricsAndAttributeNoPrimaries,
);

export const bubbleChartWith1Metric = legacyRecordedDataFacade(rec.BubbleChartWith1Metric);

export const bubbleChartWith3MetricsAndAttribute = legacyRecordedDataFacade(
    rec.BubbleChartWith3MetricsAndAttribute,
);
export const bubbleChartWith3MetricsAndAttributeNullsInData = legacyRecordedDataFacade(
    rec.BubbleChartWith3MetricsAndAttributeNullsInData,
);

//
// Combo chart
//
export const comboWithTwoMeasuresAndViewByAttribute = legacyRecordedDataFacade(
    rec.ComboChartWithTwoMeasuresViewByAttribute,
);

export const comboChartWithTwoMeasuresViewByAttributeNoBuckets = legacyRecordedDataFacade(
    rec.ComboChartWithTwoMeasuresViewByAttributeNoBuckets,
);

export const comboChartWithTwoMeasuresViewByAttributePercformat = legacyRecordedDataFacade(
    rec.ComboChartWithTwoMeasuresViewByAttributePercformat,
);

//
//
//
export const headlineWithOneMeasure = legacyRecordedDataFacade(rec.HeadlineWithOneMeasure);
export const headlineWithOneMeasureWithIdentifier = legacyRecordedDataFacade(
    rec.HeadlineWithOneMeasureWithIdentifier,
);
export const headlineWithTwoMeasures = legacyRecordedDataFacade(rec.HeadlineWithTwoMeasures);
export const headlineWithTwoMeasuresWithIdentifier = legacyRecordedDataFacade(
    rec.HeadlineWithTwoMeasuresWithIdentifier,
);
export const headlineWithTwoMeasuresFirstEmpty = legacyRecordedDataFacade(
    rec.HeadlineWithTwoMeasuresFirstEmpty,
);
export const headlineWithTwoMeasuresSecondEmpty = legacyRecordedDataFacade(
    rec.HeadlineWithTwoMeasuresSecondEmpty,
);
export const headlineWithTwoMeasuresBothEmpty = legacyRecordedDataFacade(
    rec.HeadlineWithTwoMeasuresBothEmpty,
);
export const headlineWithTwoMeasuresBothZero = legacyRecordedDataFacade(rec.HeadlineWithTwoMeasuresBothZero);
export const headlineWithTwoMeasuresFirstZero = legacyRecordedDataFacade(
    rec.HeadlineWithTwoMeasuresFirstZero,
);
export const headlineWithTwoMeasuresSecondZero = legacyRecordedDataFacade(
    rec.HeadlineWithTwoMeasuresSecondZero,
);
export const headlineWithTwoMeasuresBothSame = legacyRecordedDataFacade(rec.HeadlineWithTwoMeasuresBothSame);

//
// Heatmap recordings
//

export const heatMapWithEmptyCells = legacyRecordedDataFacade(rec.HeatMapWithEmptyCells);
export const heatMapWithMetricRowColumn = legacyRecordedDataFacade(rec.HeatMapWithMetricRowColumn);
//
// Pie chart recordings
//

export const pieChartWithMetricsOnly = legacyRecordedDataFacade(rec.PieChartWithMetricsOnly);
export const pieChartWithMetricsOnlyFundata = legacyRecordedDataFacade(rec.PieChartWithMetricsOnlyFundata);

//
// Scatter plot recording
//

export const scatterPlotWith2MetricsAndAttributeNullsInData = legacyRecordedDataFacade(
    rec.ScatterPlotWith2MetricsAndAttributeNullsInData,
);
export const scatterPlotWith2MetricsAndAttributeWithPrimary = legacyRecordedDataFacade(
    rec.ScatterPlotWith2MetricsAndAttributeWithPrimary,
);

//
// Treemap recordings
//

export const treemapWithMetricAndStackByAttribute = legacyRecordedDataFacade(
    rec.TreemapWithMetricAndStackByAttribute,
);
export const treemapWithMetricAndViewByAttribute = legacyRecordedDataFacade(
    rec.TreemapWithMetricAndViewByAttribute,
);
export const treemapWithTwoMetricsAndStackByAttribute = legacyRecordedDataFacade(
    rec.TreemapWithTwoMetricsAndStackByAttribute,
);
export const treemapWithMetricViewByAndStackByAttribute = legacyRecordedDataFacade(
    rec.TreemapWithMetricViewByAndStackByAttribute,
);
