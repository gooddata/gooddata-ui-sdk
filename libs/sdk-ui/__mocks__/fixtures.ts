// (C) 2007-2019 GoodData Corporation
import { LegacyExecutionRecording, legacyRecordedDataView } from "@gooddata/sdk-backend-mockingbird";
import { IInsight } from "@gooddata/sdk-model";
import { DataViewFacade } from "../src/base/results/facade";

export const testWorkspace = "testWorkspace";

function legacyRecordedDataFacade(rec: LegacyExecutionRecording): DataViewFacade {
    return DataViewFacade.for(legacyRecordedDataView(rec));
}

const BarChartForDrillTests = {
    definition: require("./recordings/bar_chart_for_drill_tests/definition.json"),
    response: require("./recordings/bar_chart_for_drill_tests/response.json"),
    result: require("./recordings/bar_chart_for_drill_tests/result.json"),
};

export const barChartForDrillTests = legacyRecordedDataFacade(BarChartForDrillTests);

export const insightWithPoP: IInsight = {
    insight: {
        identifier: "popMeasureInsight",
        uri: "/some/uri/popMeasureInsight",
        visualizationUrl: "local:test",
        buckets: [
            {
                localIdentifier: "measures",
                items: [
                    {
                        measure: {
                            localIdentifier: "m1",
                            title: "# Accounts with AD Query",
                            definition: {
                                measureDefinition: {
                                    item: {
                                        uri: "/gdc/md/myproject/obj/8172",
                                    },
                                },
                            },
                        },
                    },
                    {
                        measure: {
                            localIdentifier: "m1_pop",
                            definition: {
                                popMeasureDefinition: {
                                    measureIdentifier: "m1",
                                    popAttribute: {
                                        uri: "/gdc/md/myproject/obj/1514",
                                    },
                                },
                            },
                        },
                    },
                    {
                        measure: {
                            localIdentifier: "m1_previous_period",
                            definition: {
                                previousPeriodMeasure: {
                                    measureIdentifier: "m1",
                                    dateDataSets: [
                                        {
                                            dataSet: {
                                                uri: "/gdc/md/myproject/obj/921",
                                            },
                                            periodsAgo: 1,
                                        },
                                    ],
                                },
                            },
                        },
                    },
                ],
            },
        ],
        filters: [],
        title: "PoP measure",
        sorts: [],
        properties: {},
    },
};

export const insightWithPoPAndAlias: IInsight = {
    insight: {
        identifier: "popMeasureWithAliasInsight",
        uri: "/some/uri/popMeasureInsight",
        visualizationUrl: "local:test",
        buckets: [
            {
                localIdentifier: "measures",
                items: [
                    {
                        measure: {
                            localIdentifier: "m1",
                            title: "# Accounts with AD Query",
                            alias: "AD Queries",
                            definition: {
                                measureDefinition: {
                                    item: {
                                        uri: "/gdc/md/myproject/obj/8172",
                                    },
                                },
                            },
                        },
                    },
                    {
                        measure: {
                            localIdentifier: "m1_pop",
                            definition: {
                                popMeasureDefinition: {
                                    measureIdentifier: "m1",
                                    popAttribute: {
                                        uri: "/gdc/md/myproject/obj/1514",
                                    },
                                },
                            },
                        },
                    },
                    {
                        measure: {
                            localIdentifier: "m1_previous_period",
                            definition: {
                                previousPeriodMeasure: {
                                    measureIdentifier: "m1",
                                    dateDataSets: [
                                        {
                                            dataSet: {
                                                uri: "/gdc/md/myproject/obj/921",
                                            },
                                            periodsAgo: 1,
                                        },
                                    ],
                                },
                            },
                        },
                    },
                ],
            },
        ],
        filters: [],
        title: "PoP measure with alias",
        sorts: [],
        properties: {},
    },
};

export const insightWithArithmeticAndDerivedMeasures: IInsight = {
    insight: {
        identifier: "arithmeticAndDerivedMeasureInsight",
        uri: "/some/uri/arithmeticAndDerivedMeasureInsight",
        visualizationUrl: "local:test",
        buckets: [
            {
                localIdentifier: "measures",
                items: [
                    {
                        measure: {
                            localIdentifier: "am1",
                            definition: {
                                arithmeticMeasure: {
                                    operator: "sum",
                                    measureIdentifiers: ["m1", "m1_pop"],
                                },
                            },
                            title: "ignored title",
                        },
                    },
                    {
                        measure: {
                            localIdentifier: "m1",
                            title: "AD Accounts",
                            definition: {
                                measureDefinition: {
                                    item: {
                                        uri: "/gdc/md/myproject/obj/8172",
                                    },
                                },
                            },
                        },
                    },
                    {
                        measure: {
                            localIdentifier: "m1_pop",
                            definition: {
                                popMeasureDefinition: {
                                    measureIdentifier: "m1",
                                    popAttribute: {
                                        uri: "/gdc/md/myproject/obj/1514",
                                    },
                                },
                            },
                            title: "ignored title",
                        },
                    },
                ],
            },
        ],
        filters: [],
        title: "Arithmetic and derived measures",
        sorts: [],
        properties: {},
    },
};

export const insightWithArithmeticMeasureTree: IInsight = {
    insight: {
        identifier: "arithmeticMeasureTreeInsight",
        uri: "/some/uri/arithmeticMeasureTreeInsight",
        visualizationUrl: "local:test",
        buckets: [
            {
                localIdentifier: "measures",
                items: [
                    {
                        measure: {
                            localIdentifier:
                                "arithmetic_measure_created_from_complicated_arithmetic_measures",
                            definition: {
                                arithmeticMeasure: {
                                    operator: "sum",
                                    measureIdentifiers: [
                                        "arithmetic_measure_created_from_arithmetic_measures",
                                        "arithmetic_measure_created_from_simple_measures",
                                    ],
                                },
                            },
                        },
                    },
                    {
                        measure: {
                            localIdentifier: "m1",
                            title: "AD Accounts",
                            definition: {
                                measureDefinition: {
                                    item: {
                                        uri: "/gdc/md/myproject/obj/8172",
                                    },
                                },
                            },
                        },
                    },
                    {
                        measure: {
                            localIdentifier: "m2",
                            title: "KD Accounts",
                            definition: {
                                measureDefinition: {
                                    item: {
                                        uri: "/gdc/md/myproject/obj/1245",
                                    },
                                },
                            },
                        },
                    },
                    {
                        measure: {
                            localIdentifier: "tree_level_2",
                            definition: {
                                arithmeticMeasure: {
                                    operator: "sum",
                                    measureIdentifiers: ["m1", "tree_level_1"],
                                },
                            },
                        },
                    },
                    {
                        measure: {
                            localIdentifier: "tree_level_1",
                            definition: {
                                arithmeticMeasure: {
                                    operator: "sum",
                                    measureIdentifiers: ["m1", "tree_level_0"],
                                },
                            },
                        },
                    },
                    {
                        measure: {
                            localIdentifier: "tree_level_0",
                            definition: {
                                arithmeticMeasure: {
                                    operator: "sum",
                                    measureIdentifiers: ["m1", "m2"],
                                },
                            },
                        },
                    },
                ],
            },
        ],
        filters: [],
        title: "Arithmetic measure tree",
        sorts: [],
        properties: {},
    },
};

export const insightWithComplexArithmeticMeasureTree: IInsight = {
    insight: {
        identifier: "complexArithmeticMeasureTreeInsight",
        uri: "/some/uri/complexArithmeticMeasureTreeInsight",
        visualizationUrl: "local:test",
        buckets: [
            {
                localIdentifier: "measures",
                items: [
                    {
                        measure: {
                            localIdentifier:
                                "arithmetic_measure_created_from_complicated_arithmetic_measures",
                            definition: {
                                arithmeticMeasure: {
                                    operator: "sum",
                                    measureIdentifiers: [
                                        "arithmetic_measure_created_from_arithmetic_measures",
                                        "arithmetic_measure_created_from_simple_measures",
                                    ],
                                },
                            },
                        },
                    },
                    {
                        measure: {
                            localIdentifier: "m1",
                            title: "# Accounts with AD Query",
                            definition: {
                                measureDefinition: {
                                    item: {
                                        uri: "/gdc/md/myproject/obj/8172",
                                    },
                                },
                            },
                        },
                    },
                    {
                        measure: {
                            localIdentifier: "m2",
                            title: "# Accounts with KD Query",
                            definition: {
                                measureDefinition: {
                                    item: {
                                        uri: "/gdc/md/myproject/obj/1245",
                                    },
                                },
                            },
                        },
                    },
                    {
                        measure: {
                            localIdentifier: "m3",
                            title: "# Accounts with AD Query",
                            alias: "AD Queries",
                            definition: {
                                measureDefinition: {
                                    item: {
                                        uri: "/gdc/md/myproject/obj/8172",
                                    },
                                },
                            },
                        },
                    },
                    {
                        measure: {
                            localIdentifier: "m4",
                            title: "# Accounts with KD Query",
                            alias: "KD Queries",
                            definition: {
                                measureDefinition: {
                                    item: {
                                        uri: "/gdc/md/myproject/obj/1245",
                                    },
                                },
                            },
                        },
                    },
                    {
                        measure: {
                            localIdentifier: "m1_pop",
                            title: "M1 Pop Measure Title",
                            definition: {
                                popMeasureDefinition: {
                                    measureIdentifier: "m1",
                                    popAttribute: {
                                        uri: "/gdc/md/myproject/obj/1514",
                                    },
                                },
                            },
                        },
                    },
                    {
                        measure: {
                            localIdentifier: "m1_previous_period",
                            title: "M1 Previous Measure Title",
                            definition: {
                                previousPeriodMeasure: {
                                    measureIdentifier: "m1",
                                    dateDataSets: [
                                        {
                                            dataSet: {
                                                uri: "/gdc/md/myproject/obj/921",
                                            },
                                            periodsAgo: 1,
                                        },
                                    ],
                                },
                            },
                        },
                    },
                    {
                        measure: {
                            localIdentifier: "m1_pop_renamed",
                            definition: {
                                popMeasureDefinition: {
                                    measureIdentifier: "m1",
                                    popAttribute: {
                                        uri: "/gdc/md/myproject/obj/1514",
                                    },
                                },
                            },
                            alias: "Renamed SP last year M1",
                        },
                    },
                    {
                        measure: {
                            localIdentifier: "m1_previous_period_renamed",
                            definition: {
                                previousPeriodMeasure: {
                                    measureIdentifier: "m1",
                                    dateDataSets: [
                                        {
                                            dataSet: {
                                                uri: "/gdc/md/myproject/obj/921",
                                            },
                                            periodsAgo: 1,
                                        },
                                    ],
                                },
                            },
                            alias: "Renamed previous period M1",
                        },
                    },
                    {
                        measure: {
                            localIdentifier: "derived_measure_from_arithmetic_measure",
                            definition: {
                                popMeasureDefinition: {
                                    measureIdentifier: "arithmetic_measure_created_from_simple_measures",
                                    popAttribute: {
                                        uri: "/gdc/md/myproject/obj/1514",
                                    },
                                },
                            },
                        },
                    },
                    {
                        measure: {
                            localIdentifier: "arithmetic_measure_created_from_simple_measures",
                            title: "Arithmetic sum(m1, m2) Measure Title",
                            definition: {
                                arithmeticMeasure: {
                                    operator: "sum",
                                    measureIdentifiers: ["m1", "m2"],
                                },
                            },
                        },
                    },
                    {
                        measure: {
                            localIdentifier: "arithmetic_measure_created_from_renamed_simple_measures",
                            title: "Arithmetic sum(m3, m4) Measure Title",
                            definition: {
                                arithmeticMeasure: {
                                    operator: "sum",
                                    measureIdentifiers: ["m3", "m4"],
                                },
                            },
                        },
                    },
                    {
                        measure: {
                            localIdentifier: "arithmetic_measure_created_from_derived_measures",
                            definition: {
                                arithmeticMeasure: {
                                    operator: "sum",
                                    measureIdentifiers: ["m1_pop", "m1_previous_period"],
                                },
                            },
                        },
                    },
                    {
                        measure: {
                            localIdentifier: "arithmetic_measure_created_from_arithmetic_measures",
                            definition: {
                                arithmeticMeasure: {
                                    operator: "sum",
                                    measureIdentifiers: [
                                        "arithmetic_measure_created_from_renamed_simple_measures",
                                        "arithmetic_measure_created_from_renamed_derived_measures",
                                    ],
                                },
                            },
                        },
                    },
                    {
                        measure: {
                            localIdentifier: "arithmetic_measure_created_from_renamed_derived_measures",
                            definition: {
                                arithmeticMeasure: {
                                    operator: "sum",
                                    measureIdentifiers: ["m1_pop_renamed", "m1_previous_period_renamed"],
                                },
                            },
                        },
                    },
                    {
                        measure: {
                            localIdentifier: "invalid_arithmetic_measure_with_missing_dependency",
                            definition: {
                                arithmeticMeasure: {
                                    operator: "sum",
                                    measureIdentifiers: ["m1", "m666"],
                                },
                            },
                        },
                    },
                    {
                        measure: {
                            localIdentifier: "invalid_arithmetic_measure_with_cyclic_dependency_1",
                            definition: {
                                arithmeticMeasure: {
                                    operator: "sum",
                                    measureIdentifiers: [
                                        "m1",
                                        "invalid_arithmetic_measure_with_cyclic_dependency_2",
                                    ],
                                },
                            },
                        },
                    },
                    {
                        measure: {
                            localIdentifier: "invalid_arithmetic_measure_with_cyclic_dependency_2",
                            definition: {
                                arithmeticMeasure: {
                                    operator: "sum",
                                    measureIdentifiers: [
                                        "m2",
                                        "invalid_arithmetic_measure_with_cyclic_dependency_1",
                                    ],
                                },
                            },
                        },
                    },
                ],
            },
        ],
        filters: [],
        title: "Complex arithmetic measure tree",
        sorts: [],
        properties: {},
    },
};

export const insightWithMultipleMeasureBuckets: IInsight = {
    insight: {
        identifier: "multipleMeasureBucketsInsight",
        uri: "/some/uri/multipleMeasureBucketsInsight",
        visualizationUrl: "local:test",
        buckets: [
            {
                localIdentifier: "measures",
                items: [
                    {
                        measure: {
                            localIdentifier: "fdd41e4ca6224cd2b5ecce15fdabf062",
                            format: "#,##0.00",
                            title: "Sum of Email Clicks",
                            definition: {
                                measureDefinition: {
                                    aggregation: "sum",
                                    item: {
                                        uri: "/gdc/md/yrungi0zwpoud7h1kmh6ldhp0vgkpi41/obj/15428",
                                    },
                                },
                            },
                        },
                    },
                ],
            },
            {
                localIdentifier: "secondary_measures",
                items: [
                    {
                        measure: {
                            localIdentifier: "fdd41e4ca6224cd2b5ecce15fdabf062_pop",
                            definition: {
                                popMeasureDefinition: {
                                    measureIdentifier: "fdd41e4ca6224cd2b5ecce15fdabf062",
                                    popAttribute: {
                                        uri: "/gdc/md/yrungi0zwpoud7h1kmh6ldhp0vgkpi41/obj/15330",
                                    },
                                },
                            },
                        },
                    },
                    {
                        measure: {
                            localIdentifier: "fdd41e4ca6224cd2b5ecce15fdabf062_previous_period",
                            definition: {
                                previousPeriodMeasure: {
                                    measureIdentifier: "fdd41e4ca6224cd2b5ecce15fdabf062",
                                    dateDataSets: [
                                        {
                                            dataSet: {
                                                uri: "/gdc/md/yrungi0zwpoud7h1kmh6ldhp0vgkpi41/obj/921",
                                            },
                                            periodsAgo: 1,
                                        },
                                    ],
                                },
                            },
                        },
                    },
                ],
            },
        ],
        filters: [],
        title: "Multiple measure buckets",
        sorts: [],
        properties: {},
    },
};
