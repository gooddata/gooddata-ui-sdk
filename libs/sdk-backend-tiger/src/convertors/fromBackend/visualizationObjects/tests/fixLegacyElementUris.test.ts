// (C) 2019-2022 GoodData Corporation
import { IInsightWidget } from "@gooddata/sdk-model";
import {
    ColorMapping,
    fixInsightLegacyElementUris,
    fixWidgetLegacyElementUris,
} from "../../fixLegacyElementUris.js";
import { mockInsight, mockWidget } from "./fixLegacyElementUris.fixtures.js";
import { describe, expect, it } from "vitest";

describe("fixInsightLegacyElementUris", () => {
    const insightWithUris = mockInsight([
        "/obj/0/elements?id=CZ",
        "/obj/0/elements?id=Czech.Republic",
        "/obj/0/elements?id=Czech Republic",
        "/obj/9999999999/elements?id=CZ",
    ]);
    const insightWithoutUris = mockInsight(["", "Untouched", "Untouched.string", "Should be untouched"]);
    const sanitizedInsightWithUris = fixInsightLegacyElementUris(insightWithUris);
    const sanitizedInsightWithoutUris = fixInsightLegacyElementUris(insightWithoutUris);

    describe("color mapping", () => {
        it("should strip fake uris and keep only label text values", () => {
            expect(sanitizedInsightWithUris.insight.properties.controls?.colorMapping).toMatchSnapshot();
        });

        it("should not change elements with correct primary key", () => {
            expect(insightWithoutUris.insight.properties.controls?.colorMapping).toEqual(
                sanitizedInsightWithoutUris.insight.properties.controls?.colorMapping,
            );
        });

        it("should handle color mapping items with null ids (RAIL-3983)", () => {
            const fakeInsight = mockInsight(["/obj/0/elements?id=CZ", "/obj/0/elements?id=Czech.Republic"]);
            const emptyValueColorMapping: ColorMapping = {
                id: null, // set this color for the (empty value) attribute value
                color: {
                    type: "guid",
                    value: "foo",
                },
            };
            fakeInsight.insight.properties.controls.colorMapping.push(emptyValueColorMapping);

            expect(
                fixInsightLegacyElementUris(fakeInsight).insight.properties.controls.colorMapping,
            ).toMatchSnapshot();
        });
    });

    describe("column widths", () => {
        it("should strip fake uris and keep only label text values", () => {
            expect(sanitizedInsightWithUris.insight.properties.controls?.columnWidths).toMatchSnapshot();
        });

        it("should not change elements with correct primary key", () => {
            expect(insightWithoutUris.insight.properties.controls?.columnWidths).toEqual(
                sanitizedInsightWithoutUris.insight.properties.controls?.columnWidths,
            );
        });
    });

    describe("sort items", () => {
        it("should strip fake uris and keep only label text values", () => {
            expect(sanitizedInsightWithUris.insight.sorts).toMatchSnapshot();
        });

        it("should copy sort items to insight properties", () => {
            expect(sanitizedInsightWithUris.insight.sorts).toEqual(
                sanitizedInsightWithUris.insight.properties.sortItems,
            );
        });

        it("should not change elements with correct primary key", () => {
            expect(insightWithoutUris.insight.sorts).toEqual(
                sanitizedInsightWithoutUris.insight.properties?.sortItems,
            );

            expect(insightWithoutUris.insight.sorts).toEqual(sanitizedInsightWithoutUris.insight.sorts);
        });
    });
});

describe("fixWidgetLegacyElementUris", () => {
    const widgetWithUris = mockWidget([
        "/obj/0/elements?id=CZ",
        "/obj/0/elements?id=Czech.Republic",
        "/obj/0/elements?id=Czech Republic",
        "/obj/9999999999/elements?id=CZ",
    ]);
    const widgetWithoutUris = mockWidget([
        "",
        "Untouched",
        "Untouched.string",
        "Should be untouched",
    ]) as IInsightWidget;
    const sanitizedWidgetWithUris = fixWidgetLegacyElementUris(widgetWithUris) as IInsightWidget;
    const sanitizedWidgetWithoutUris = fixWidgetLegacyElementUris(widgetWithoutUris) as IInsightWidget;

    describe("column widths", () => {
        it("should strip fake uris and keep only label text values", () => {
            expect(sanitizedWidgetWithUris.properties?.controls?.columnWidths).toMatchSnapshot();
        });

        it("should not change elements with correct primary key", () => {
            expect(widgetWithoutUris.properties?.controls?.columnWidths).toEqual(
                sanitizedWidgetWithoutUris.properties?.controls?.columnWidths,
            );
        });
    });
});
