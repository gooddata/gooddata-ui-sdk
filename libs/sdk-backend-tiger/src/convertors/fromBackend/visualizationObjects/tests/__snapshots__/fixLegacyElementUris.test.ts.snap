// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fixInsightLegacyElementUris color mapping should handle color mapping items with null ids (RAIL-3983) 1`] = `
Array [
  Object {
    "color": Object {
      "type": "guid",
      "value": "0",
    },
    "id": "CZ",
  },
  Object {
    "color": Object {
      "type": "guid",
      "value": "1",
    },
    "id": "Czech.Republic",
  },
  Object {
    "color": Object {
      "type": "guid",
      "value": "foo",
    },
    "id": null,
  },
]
`;

exports[`fixInsightLegacyElementUris color mapping should strip fake uris and keep only label text values 1`] = `
Array [
  Object {
    "color": Object {
      "type": "guid",
      "value": "0",
    },
    "id": "CZ",
  },
  Object {
    "color": Object {
      "type": "guid",
      "value": "1",
    },
    "id": "Czech.Republic",
  },
  Object {
    "color": Object {
      "type": "guid",
      "value": "2",
    },
    "id": "Czech Republic",
  },
  Object {
    "color": Object {
      "type": "guid",
      "value": "3",
    },
    "id": "CZ",
  },
]
`;

exports[`fixInsightLegacyElementUris column widths should strip fake uris and keep only label text values 1`] = `
Array [
  Object {
    "measureColumnWidthItem": Object {
      "locators": Array [
        Object {
          "attributeLocatorItem": Object {
            "element": "CZ",
          },
        },
      ],
    },
  },
  Object {
    "measureColumnWidthItem": Object {
      "locators": Array [
        Object {
          "attributeLocatorItem": Object {
            "element": "Czech.Republic",
          },
        },
      ],
    },
  },
  Object {
    "measureColumnWidthItem": Object {
      "locators": Array [
        Object {
          "attributeLocatorItem": Object {
            "element": "Czech Republic",
          },
        },
      ],
    },
  },
  Object {
    "measureColumnWidthItem": Object {
      "locators": Array [
        Object {
          "attributeLocatorItem": Object {
            "element": "CZ",
          },
        },
      ],
    },
  },
]
`;

exports[`fixInsightLegacyElementUris sort items should strip fake uris and keep only label text values 1`] = `
Array [
  Object {
    "measureSortItem": Object {
      "direction": "asc",
      "locators": Array [
        Object {
          "attributeLocatorItem": Object {
            "attributeIdentifier": "localAttributeId",
            "element": "CZ",
          },
        },
      ],
    },
  },
  Object {
    "measureSortItem": Object {
      "direction": "asc",
      "locators": Array [
        Object {
          "attributeLocatorItem": Object {
            "attributeIdentifier": "localAttributeId",
            "element": "Czech.Republic",
          },
        },
      ],
    },
  },
  Object {
    "measureSortItem": Object {
      "direction": "asc",
      "locators": Array [
        Object {
          "attributeLocatorItem": Object {
            "attributeIdentifier": "localAttributeId",
            "element": "Czech Republic",
          },
        },
      ],
    },
  },
  Object {
    "measureSortItem": Object {
      "direction": "asc",
      "locators": Array [
        Object {
          "attributeLocatorItem": Object {
            "attributeIdentifier": "localAttributeId",
            "element": "CZ",
          },
        },
      ],
    },
  },
]
`;

exports[`fixWidgetLegacyElementUris column widths should strip fake uris and keep only label text values 1`] = `
Array [
  Object {
    "measureColumnWidthItem": Object {
      "locators": Array [
        Object {
          "attributeLocatorItem": Object {
            "element": "CZ",
          },
        },
      ],
    },
  },
  Object {
    "measureColumnWidthItem": Object {
      "locators": Array [
        Object {
          "attributeLocatorItem": Object {
            "element": "Czech.Republic",
          },
        },
      ],
    },
  },
  Object {
    "measureColumnWidthItem": Object {
      "locators": Array [
        Object {
          "attributeLocatorItem": Object {
            "element": "Czech Republic",
          },
        },
      ],
    },
  },
  Object {
    "measureColumnWidthItem": Object {
      "locators": Array [
        Object {
          "attributeLocatorItem": Object {
            "element": "CZ",
          },
        },
      ],
    },
  },
]
`;
